src/lib/formatters.js

```diff
@@ -112,6 +112,18 @@ export function formatForDisplay(amount, fix) {
 
 }
 
+export function formatCAPForDisplay(amountStr) {
```

Comment:
Any reason we can't do this?

```
export function formatCAPForDisplay(amount) {
return Math.ceil(amount * 10**6) / 10**6;
}
```

Reply:
Can't do that exactly because it will only capture the dust amounts when there is nothing but dust. Also the function argument, if sourced from formatUnit, will be a string representation of the number.

Your comment got me thinking though, for any `CAP_MAX_DISPLAY_DIGITS` >= 3 we could safely convert the insignificant digits to the right of the decimal to an integer in JS and then check whether this is greater than zero to determine if there is dust to account for. That will be clearer in intent than the current version, let me update.

Reply:
@ModernForays I believe we don't have to deal with the dust amounts if we can round up, since the contracts will automatically select the max available, even if the submitted amount is above it. 

So for example:
* 0.123456789 displayed as 0.123457 will work (when submitted as 0.123457 to the contracts, the contracts will detect it is more than the max and pull the precise max available).

Which is why I thought my initial comment would make sense: https://github.com/capofficial/client/pull/41#discussion_r1044257844

In the UI the user will see their CAP amount rounded up to the 6th decimal.

To convert the string to an int before passing it to formatCAPForDisplay, we can do amountStr*1, like formatCAPForDisplay("9.03931"*1).

To make it even simpler, we can probably use the formatForDisplay method directly (granted it uses toFixed so it will truncate down on the 6th decimal but the user won't care about an extra 0.000001 CAP). I believe this is used in the Close Position modal when clicking on the position size to fill it in quickly.

Am I missing something here? Are there examples where that wouldn't work?

Reply:
Hey @kappacappa, thanks for the detailed feedback. I might be being a little too precise in how I approach this.

> Are there examples where that wouldn't work?

If I understand correctly, I think these would be the issues:

1. The user can manually enter an amount to stake/unstake but we should not round the user-entered values. If a user decides to unstake 2 of their 4 CAP stake, always adding 5e-7 and rounding up before sending the transaction would lead to them ending up with 1.999999 remaining in the stake.
2. If we always round up when a user clicks the max stake/unstake link as opposed to always rounding up when sending the transaction (to avoid the issue in case 1), it would be a bad experience for users doing a max stake/unstake with an integer value of CAP. Unconditionally adding 5e-7 and rounding here would result in the true 2.0 CAP balance displayed as 2.000001 CAP.

Case 1 demonstrates why I think the rounding up should be done only when populating max stake/unstake values, and case 2 shows why I think only rounding up when the user has these small fractions of a CAP in their wallet/stake makes sense.

> To convert the string to an int before passing it to formatCAPForDisplay, we can do amountStr*1, like formatCAPForDisplay("9.03931"*1).

My reason for preferring to operate on the string here is that it comes directly from ethers so as far as I know, has total fidelity with the user's wallet balance/staking amount. I wanted to avoid cases both where due to floating point error the staked amount lost precision or was shown with additional precision due to rounding error.

```javascript
> "0.33333333333333335" * 1
0.33333333333333337

> "0.100000000000000009" * 1
0.1
```

> To make it even simpler, we can probably use the formatForDisplay method directly (granted it uses toFixed so it will truncate down on the 6th decimal but the user won't care about an extra 0.000001 CAP).

I experimented with `formatForDisplay` and can change it back. One big difference between it and `formatCAPForDisplay` is that the latter uses ethers.FixedNumber's toString method, which will truncate trailing zeroes. In testing, making the max stake operation for a wallet balance of 1200 CAP populate the input with "12.0" (truncating zeroes) instead of "12.000000" while still allowing for "12.123456" -> "12.123456" felt preferable to me.

Reply:
Thanks @ModernForays - some comments:

> 1. The user can manually enter an amount to stake/unstake but we should not round the user-entered values. If a user decides to unstake 2 of their 4 CAP stake, always adding 5e-7 and rounding up before sending the transaction would lead to them ending up with 1.999999 remaining in the stake.

Correct, amounts entered manually by the user shouldn't use the formatting method. It's used just in the LabelValue or for the input value when the value is clicked from the LabelValue.

> 2. If we always round up when a user clicks the max stake/unstake link as opposed to always rounding up when sending the transaction (to avoid the issue in case 1), it would be a bad experience for users doing a max stake/unstake with an integer value of CAP. Unconditionally adding 5e-7 and rounding here would result in the true 2.0 CAP balance displayed as 2.000001 CAP.

The rounding up is not on an integer basis, but on the 6th decimal. For example, with a method like this:

```
export function formatCAPForDisplay(amount) {
    if (!amount) return 0;
    if (amount < 0.000001) return 0.000001;
    return Math.ceil(amount * 10**6) / 10**6;
}
```

we get

```
formatCAPForDisplay(5) -> 5
formatCAPForDisplay(1.23456789) -> 1.23457
formatCAPForDisplay(0.0001) -> 0.0001
formatCAPForDisplay(0.0000000000000001) -> 0.000001

```

These return values seem OK to display and the contracts will deal with them appropriately even if the amount submitted is above the max available for the user.

> My reason for preferring to operate on the string here is that it comes directly from ethers so as far as I know, has total fidelity with the user's wallet balance/staking amount. I wanted to avoid cases both where due to floating point error the staked amount lost precision or was shown with additional precision due to rounding error.

I don't believe that `Math.ceil(amount * 10**6) / 10**6` can return a number with a floating precision issue (like 1.2300000000000000001). I might be wrong but I think I remember divisions by 10x multipliers of integers, it doesn't happen, since JavaScript can convert that fraction to a precise decimal in all cases. In a console I can't seem to get a floating point error by dividing an int by a multiple of 10.

Reply:
My hang up here has been based on the general pitfalls of handling real numbers as floats (e.g. `0.1` actually being `1.00000000000000005551115123126e-1`) when accuracy is desired:

```javascript
> let a = "0.01" * 1;     // good
> Math.ceil(a * 10**2) / 10**2;
0.01
> let b = "1001.01" * 1;  // bad
> Math.ceil(b * 10**23) / 10**23;
1001.0099999999999
> let c = 0.1 / 10;       // good
> Math.ceil(c * 10**2) / 10**2;
0.01
> let d = 0.1 * 0.1;      // bad
> Math.ceil(d * 10**2) / 10**2;
0.02
```

Considering your points very carefully, I think I agree that with this specific expression, with the specific amounts we are passing to it, the specific powers of 10 involved, and the accuracy we need, we shouldn't see issues. Thanks for bearing with me on this. Will change :+1:.

Reply:
Updated behavior for `formatCAPForDisplay` below. The cases where behavior doesn't match due to use of floating point values versus `ethers.FixedNumber` are inconsequential.

| | Input | Actual | Expected |
| --- | --- | --- | --- |
| :heavy_check_mark: | "0" | 0 | 0 |
| :heavy_check_mark: | "1" | 1 | 1 |
| :heavy_check_mark: | "0.1000" | 0.1 | 0.1 |
| :heavy_check_mark: | "1001.01" | 1001.01 | 1001.01 |
| :heavy_check_mark: | "1001.010000005" | 1001.010001 | 1001.010001 |
| :heavy_check_mark: | "0.000001" | 0.000001 | 0.000001 |
| :heavy_check_mark: | "0.0000001" | 0.000001 | 0.000001 |
| :heavy_check_mark: | "0.1000001" | 0.100001 | 0.100001 |
| :x: | "540.000000000000000001" | 540 | 540.000001 |
| :x: | "540.00000000000000001" | 540 | 540.000001 |
| :x: | "540.0000000000000001" | 540 | 540.000001 |
| :x: | "540.000000000000001" | 540 | 540.000001 |
| :x: | "540.00000000000001" | 540 | 540.000001 |
| :heavy_check_mark: | "540.0000000000001" | 540.000001 | 540.000001 |
| :heavy_check_mark: | "540.000000000001" | 540.000001 | 540.000001 |
| :heavy_check_mark: | "540.00000000001" | 540.000001 | 540.000001 |
| :heavy_check_mark: | "540.0000000001" | 540.000001 | 540.000001 |
| :heavy_check_mark: | "540.000000001" | 540.000001 | 540.000001 |
| :heavy_check_mark: | "540.00000001" | 540.000001 | 540.000001 |